crypto/src/keys.rs

```diff
@@ -121,15 +125,44 @@ pub struct ProofAuthorizationKey {
     pub nsk: NullifierPrivateKey,
 }
 
+impl ProofAuthorizationKey {
+    pub fn derive(ask: &SigningKey<SpendAuth>, nsk: &NullifierPrivateKey) -> Self {
+        Self {
+            ak: ask.into(),
+            nsk: *nsk,
+        }
+    }
+}
+
 /// The `FullViewingKey` allows one to identify incoming and outgoing notes only.
 pub struct FullViewingKey {
     pub ak: VerificationKey<SpendAuth>,
     pub nk: NullifierDerivingKey,
     pub ovk: OutgoingViewingKey,
 }
 
+impl FullViewingKey {
+    pub fn derive(
+        ak: &VerificationKey<SpendAuth>,
+        nsk: &NullifierPrivateKey,
+        ovk: &OutgoingViewingKey,
```

Comment:
similar question as above about references/ownership.

Is there a misuse risk from taking input parameters like this -- what happens if the inputs that are passed in aren't all derived from the same root key material? Could we alternately have the FVK derived from another structure that would ensure all the key material is consistent?